### PR TITLE
implement Hash for various types and replace PartialOrd

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

Closes #.

 # Rationale for this change

Cherry-pick some changes from https://github.com/apache/arrow-datafusion/pull/1556 to reduce future merge overhead

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by houqp — APPROVED** [review]
Thank you @Jimexist for porting this!

**Comment by houqp** [comment]
For the context, this is because arrow2's datatype doesn't derive PartialOrd anymore because order between types is not defined semantically.

File: datafusion/src/logical_plan/expr.rs [hunk 1]
```diff
@@ -372,6 +373,23 @@ pub enum Expr {
     Wildcard,
 }
 
+/// Fixed seed for the hashing so that Ords are consistent across runs
```
Comment (alamb) [inline datafusion/src/logical_plan/expr.rs:376]: 👍

File: datafusion/src/logical_plan/window_frames.rs [hunk 2]
```diff
@@ -190,6 +191,12 @@ impl Ord for WindowFrameBound {
     }
 }
 
+impl Hash for WindowFrameBound {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.get_rank().hash(state)
```
Comment (alamb) [inline datafusion/src/logical_plan/window_frames.rs:196]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me

**Pull request merged**